### PR TITLE
Feature Incremental Sort

### PR DESCRIPTION
**Pull request by wangxuqi**
Backport Implement Incremental Sort from PG14.

Incremental Sort is an optimized variant of multikey sort for cases when the input is already sorted by a prefix of the requested sort keys. For example when the relation is already sorted by (key1, key2) and we need to sort it by (key1, key2, key3) we can simply split the input rows into groups having equal values in (key1, key2), and only sort/compare the remaining column key3.

This has a number of benefits:

- Reduced memory consumption, because only a single group (determined by   values in the sorted prefix) needs to be kept in memory. This may also   eliminate the need to spill to disk.

- Lower startup cost, because Incremental Sort produce results after each   prefix group, which is beneficial for plans where startup cost matters   (like for example queries with LIMIT clause).

We consider both Sort and Incremental Sort, and decide based on costing.

The implemented algorithm operates in two different modes:

- Fetching a minimum number of tuples without check of equality on the   prefix keys, and sorting on all columns when safe.

- Fetching all tuples for a single prefix group and then sorting by   comparing only the remaining (non-prefix) keys.

We always start in the first mode, and employ a heuristic to switch into the second mode if we believe it's beneficial - the goal is to minimize the number of unnecessary comparions while keeping memory consumption below work_mem.

This is a very old patch series. The idea was originally proposed by Alexander Korotkov back in 2013, and then revived in 2017. In 2018 the patch was taken over by James Coleman, who wrote and rewrote most of the current code.

There were many reviewers/contributors since 2013 - I've done my best to pick the most active ones, and listed them in this commit message.

Author: James Coleman, Alexander Korotkov
Reviewed-by: Tomas Vondra, Andreas Karlsson, Marti Raudsepp, Peter Geoghegan, Robert Haas, Thomas Munro, Antonin Houska, Andres Freund, Alexander Kuzmenkov
Discussion: https://postgr.es/m/CAPpHfdscOX5an71nHd8WSUH6GNOCf=V7wgDaTXdDd9=goN-gfA@mail.gmail.com
Discussion: https://postgr.es/m/CAPpHfds1waRZ=NOmueYq0sx1ZSCnt+5QJvizT8ndT2=etZEeAQ@mail.gmail.com

### CODE REVIEW

**Comment by avamingli** [comment]
Hi, there are some subsequent Incremental Sort related commits, ex:
lacks of Incremental Sort support for Window function.
Shall we enable them all if decide to cherry-pick this feature from upstream?
And a few refactor may be needed  in such function `cdb_prepare_path_for_sorted_agg` for GPDB.

**Comment by wangxuqi** [comment]
> 
Yes , there are more work to do on window function support and so on.  But I think this base commit at least can work on some Top-N queries. Point as you mentioned above and include ORCA for  Incremental Sort will be implement in another PR.

**Comment by wangxuqi** [comment]
> 

It seems there is no need to do refactor on functions like `cdb_prepare_path_for_sorted_agg `. As for upper node, Incremental Sort Node is no different with Sort Node. `is_sorted` will be marked to 'true' same as Sort Node. See the plan below
```
explain (costs off)
select * from (select * from tenk1 order by four) t order by four, ten
    limit 1;
                     QUERY PLAN                      
-----------------------------------------------------
 Limit
   ->  Gather Motion 3:1  (slice1; segments: 3)
         Merge Key: tenk1.four, tenk1.ten
         ->  Limit
               ->  Incremental Sort
                     Sort Key: tenk1.four, tenk1.ten
                     Presorted Key: tenk1.four
                     ->  Sort
                           Sort Key: tenk1.four
                           ->  Seq Scan on tenk1
 Optimizer: Postgres query optimizer
(11 rows)
```
The result in gpdb is also sorted correctly.

**Comment by wangxuqi** [comment]
> Hi, there are some subsequent Incremental Sort related commits, ex: lacks of Incremental Sort support for Window function. Shall we enable them all if decide to cherry-pick this feature from upstream? And a few refactor may be needed in such function `cdb_prepare_path_for_sorted_agg` for GPDB.

Done

File: src/backend/cdb/cdbgroupingpaths.c [hunk 1]
```diff
@@ -2351,11 +2352,28 @@ cdb_prepare_path_for_sorted_agg(PlannerInfo *root,
...
+				 * Since we have presorted keys and incremental sort is
+				 * enabled, just use incremental sort.
+				 */
+				subpath = (Path *) create_incremental_sort_path(root,
+																rel,
+																subpath,
+																group_pathkeys,
+																presorted_keys,
+																-1.0);
+			}
```
Comment (avamingli) [inline src/backend/cdb/cdbgroupingpaths.c:2376]: `create_incremental_sort_path` could also be used below `else if (!is_sorted && group_pathkeys)` (at line 2411) in this function.

Comment (wangxuqi) [reply src/backend/cdb/cdbgroupingpaths.c:2376]: Nice suggestion. make it work

Comment (wangxuqi) [reply src/backend/cdb/cdbgroupingpaths.c:2377]: Done

**Comment by ashwinstar** [comment]
Given this is a purely performance centered feature need to provide numbers to understand the benefit cost vs value aspect from out of line backporting (as main branch is till PG12) this feature to GPDB. Also, would be helpful to provide clear details on what's the remaining work for the same.

**Comment by soumyadeep2007** [comment]
Given we haven't heard back from @wangxuqi and we don't have enough motivation to justify an out-of-line backport, I'm closing this for now. @wangxuqi, please feel free to reopen or comment here if you get a chance.

**Pull request closed without merging**